src/FSAL/FSAL_CORTXFS/mds.c

```diff
@@ -269,25 +280,146 @@ int kvsfs_pmds_ini(struct kvsfs_fsal_module *kvsfs,
+		LogCrit(COMPONENT_PNFS,"pNFS parameters configured incorrectly\n");
+		rc = -rc;
+		goto out; 
+	}
+	pNFS_role = get_string_config_value(items, NULL);
+	if(strcmp(pNFS_role,"MDS") == 0) {
+		mds_ctx->pnfs_role = KVSFS_PNFS_MDS;
+	} 
+	else if(strcmp(pNFS_role,"DS") == 0) { 
+		mds_ctx->pnfs_role = KVSFS_PNFS_DS;
+	}
+	else if(strcmp(pNFS_role,"BOTH") == 0)	{
+		mds_ctx->pnfs_role = KVSFS_PNFS_BOTH;
+	}
+	else {
+		mds_ctx->pnfs_role = KVSFS_PNFS_DISABLED;
+	}
+	
+	
+	if(mds_ctx->pnfs_role > KVSFS_PNFS_DS )
+	{	
+		mds_ctx->mds = gsh_calloc(1, sizeof(struct kvsfs_pnfs_mds_info));
+		if (mds_ctx->mds == NULL) {
+			LogCrit(COMPONENT_PNFS,
 			"gsh_calloc failed: %u", sizeof(*mds_ctx->mds));
-		rc = -ENOMEM;
-		goto out;
+			rc = -ENOMEM;
+			goto out;
+		}
+		mds_ctx->mds->stripe_unit = KVSFS_PNFS_STRIPE_UNIT;	
+	
+		rc = get_config_item("pNFS","NumDs",cfg_items, &items);
+		if ( items == NULL){
+			LogCrit(COMPONENT_PNFS,"Zero data servers\n");
+			rc = -rc; 
+			goto out;
+		}
+		
+		num_ds=(int)get_int_config_value(items, 0, 0, NULL);
+		
+		LIST_INIT(&mds_ctx->mds->mds_ds_list);
+			
+		char DS_List[256],tmp[256];
+                char *DS_Info,*port,*ipaddr;
+		int i;	
+	
+		for(i=1; i<=num_ds; i++)
+		{
+			snprintf(DS_List,sizeof(DS_List),"DS%d",i);
+			LogDebug(COMPONENT_PNFS,"DS_List is %s",DS_List);
+			rc = get_config_item("pNFS",DS_List,cfg_items, &items);
+			if ( items == NULL)
+			{
+				LogCrit(COMPONENT_PNFS,"No Data Server Entry DS%d",i);
+				rc = -rc; 
+				goto out;
+			}
+			
+			ds = gsh_calloc(1, sizeof(*ds));
+			if ( ds == NULL ) {
+				LogCrit(COMPONENT_PNFS, "gsh_calloc failed: %u", sizeof(*ds));
+				rc = -ENOMEM;
+				goto out;
+			}			
+ 			/**
+                         * TODO: Once dynamic add/remove DS support is added
+                         * either via temp. config file or via HA, we will need
+                         * to design how to update and maintain ds_id
+                         */
+                        ds->ds_id = i;
+                        LIST_INIT(&ds->ds_addr_list);
+
+                        // per DS, only one addr as of now
+                        // TODO: support for multiple addresses
+                        ds_addr = gsh_calloc(1, sizeof(*ds_addr));
+                        if (ds_addr == NULL) {
+                                LogCrit(COMPONENT_PNFS,
+                                        "gsh_calloc failed: %u",
+                                        sizeof(*ds_addr));
+                                rc = -ENOMEM;
+                                goto out;
+                        }
+                        DS_Info = get_string_config_value(items,NULL);
+			if (DS_Info == NULL){
+				LogCrit(COMPONENT_PNFS,
+					"No valid IP address or Port found\n");
+				goto out;
+			}
+			/* Get the IP address */
+			strcpy(tmp,DS_Info);
+			ipaddr = strtok(tmp,":");
+			ds_addr->ds_ip_addr.sin_addr.s_addr = inet_addr(ipaddr);	
+
+			/* Get the port number */
+			strcpy(tmp,DS_Info);
```

Comment:
We should use strncpy, same applicable for other instances.

Reply:
@pratyush-seagate  Codacy was not passing because strncpy() does not guarantee that the destination string will be NULL terminated.I agree that with strcpy , there can be overflow but Codacy analyzed that for the strings that i have defined the buffer is large enough to hold them and Codacy passes.

Reply:
With strncpy, the last byte of the dst should be externally set to '\0', that should also satisfy codacy.
e.g.
strncpy(dst, src, <size_of_dst>);
dst[<size_of_dst> - 1] = '\0';

The above two statements can be packed inside a convenient macro for cleaner code. If glibc had strlcpy(), then this problem with strncpy would not exist.

Reply:
Codacy refuses to accept strncpy.It calls it "[MS-banned]" .This is a list of  functions that are in the banned list of functions released by Microsoft.I am keeping it to strcpy with a check for the size